### PR TITLE
Make CUDA algorithms identical to SYCL algorithms

### PR DESCRIPTION
**Pull request by beomki-yeo**
Depends on #207 

This PR makes CUDA algorithms take `traccc::memory_resource` and buffer type inputs, which follows the SYCL algorithm style.
Following is the benchmark result of `seq_example_cuda` and `seq_example_sycl` (single precision / Release build / RTX 2070)

### CUDA Log

```
[beomki@localhost traccc_build]$ ./bin/traccc_seq_example_cuda --detector_file=tml_detector/trackml-detector.csv --digitization_config_file=tml_detector/default-geometric-config-generic.json   --input_directory=tml_full/ttbar_mu100/   --events=5 --run_cpu=1 
Running ./bin/traccc_seq_example_cuda tml_detector/trackml-detector.csv tml_full/ttbar_mu100/ 5
event 0
 number of seeds (cpu): 8210
 number of seeds (cuda): 8210
 measurements matching rate: 0.999581
 spacepoint matching rate: 0.968937
 seed matching rate: 0.949452
 track parameters matching rate: 0.988063
event 1
 number of seeds (cpu): 7637
 number of seeds (cuda): 7636
 measurements matching rate: 0.999565
 spacepoint matching rate: 0.971408
 seed matching rate: 0.960849
 track parameters matching rate: 0.993322
event 2
 number of seeds (cpu): 5899
 number of seeds (cuda): 5899
 measurements matching rate: 0.999707
 spacepoint matching rate: 0.969732
 seed matching rate: 0.946771
 track parameters matching rate: 0.992372
event 3
 number of seeds (cpu): 6505
 number of seeds (cuda): 6505
 measurements matching rate: 0.999749
 spacepoint matching rate: 0.970779
 seed matching rate: 0.954189
 track parameters matching rate: 0.994312
event 4
 number of seeds (cpu): 7025
 number of seeds (cuda): 7024
 measurements matching rate: 0.99966
 spacepoint matching rate: 0.971588
 seed matching rate: 0.957011
 track parameters matching rate: 0.992883
==> Statistics ... 
- read    243510 spacepoints from 55215 modules
- created        848282 cells           
- created        243510 meaurements     
- created        243510 spacepoints     
- created (cpu)  35276 seeds
- created (cuda) 35274 seeds
==> Elpased time ... 
wall time           12.5739   
file reading (cpu)        9.6889    
clusterization_time (cpu) 0.0507225 
spacepoint_formation_time (cpu) 0.00747698
clusterization and sp formation (cuda) 0.0353263 
seeding_time (cpu)        0.237097  
seeding_time (cuda)       0.0292884 
tr_par_esti_time (cpu)    0.00780452
tr_par_esti_time (cuda)   0.001805  
```

### SYCL Log

```
[beomki@localhost traccc_sycl_build]$ ./bin/traccc_seq_example_sycl --detector_file=tml_detector/trackml-detector.csv --digitization_config_file=tml_detector/default-geometric-config-generic.json   --input_directory=tml_full/ttbar_mu100/   --events=5 --run_cpu=1
Running ./bin/traccc_seq_example_sycl tml_detector/trackml-detector.csv tml_full/ttbar_mu100/ 5
Running Seeding on device: NVIDIA GeForce RTX 2070
event 0
 number of seeds (cpu): 8210
 number of seeds (sycl): 8210
 measurements matching rate: 0.999581
 spacepoint matching rate: 0.968803
 seed matching rate: 0.946529
 track parameters matching rate: 0.987698
event 1
 number of seeds (cpu): 7637
 number of seeds (sycl): 7636
 measurements matching rate: 0.999546
 spacepoint matching rate: 0.971692
 seed matching rate: 0.959146
 track parameters matching rate: 0.991882
event 2
 number of seeds (cpu): 5899
 number of seeds (sycl): 5899
 measurements matching rate: 0.99973
 spacepoint matching rate: 0.96962
 seed matching rate: 0.948805
 track parameters matching rate: 0.993558
event 3
 number of seeds (cpu): 6505
 number of seeds (sycl): 6505
 measurements matching rate: 0.999772
 spacepoint matching rate: 0.97062
 seed matching rate: 0.952806
 track parameters matching rate: 0.994927
event 4
 number of seeds (cpu): 7025
 number of seeds (sycl): 7025
 measurements matching rate: 0.99962
 spacepoint matching rate: 0.971448
 seed matching rate: 0.959004
 track parameters matching rate: 0.993737
==> Statistics ... 
- read    243510 spacepoints from 55215 modules
- created        848282 cells           
- created        243510 meaurements     
- created        243510 spacepoints     
- created (sycl) 243510 spacepoints     
- created (cpu)  35276 seeds
- created (sycl) 35275 seeds
==> Elpased time ... 
wall time           12.1112   
file reading (cpu)        9.50571   
clusterization_time (cpu) 0.0540716 
spacepoint_formation_time (cpu) 0.00780281
clusterization_time (sycl) 0.0462837 
seeding_time (cpu)        0.227384  
seeding_time (sycl)       0.0237148 
tr_par_esti_time (cpu)    0.00851142
tr_par_esti_time (sycl)   0.00196559
```

### CODE REVIEW

**Comment by beomki-yeo** [comment]
@krasznaa Can you review the PR soon? 
I am going to release the traccc version after merging this PR and generate the results to be included in the Connecting the Dots Proceeding

**Comment by stephenswat** [comment]
Hi Beomki, just for your information Attila's on a holiday right now so he might not be as expedient with his reviews as normal.

**Comment by beomki-yeo** [comment]
OK thanks for the info. @stephenswat @konradkusiak97 might be able to review the PR

**Comment by beomki-yeo** [comment]
@Chamodya-ka asked to call device-to-host memory transfer (required for performance check and comparison between cpu and cuda) only when required. So I moved them inside the if statement.
Also file reading for CPU is called only when `run_cpu=1`.

File: examples/run/cuda/seq_example_cuda.cpp [hunk 1]
```diff
@@ -268,35 +333,46 @@ int seq_run(const traccc::full_tracking_input_config& i_cfg,
         n_cells += cells_per_event.total_size();
         n_measurements += measurements_per_event.total_size();
         n_spacepoints += spacepoints_per_event.total_size();
+        n_spacepoints_cuda += spacepoints_per_event_cuda.total_size();
         n_seeds_cuda += seeds_cuda.size();
```
Comment (Chamodya-ka) [inline examples/run/cuda/seq_example_cuda.cpp:337]: Thanks for adding the change as requested, however seeds_cuda will be empty if `run_cpu=0` right? And the statistics wouldn't make sense in that case, is that okay?

Comment (beomki-yeo) [reply examples/run/cuda/seq_example_cuda.cpp]: Yeah you are right. I don't see a nice way to handle the statistics part without transferring the memory to host :/
We might need to change statistics part more ergonomically at some point.

**Review by stephenswat — APPROVED** [review]
Have not had time to look into this in depth, but if you're rushing to get this in I'm happy to sign off on it.

**Comment by beomki-yeo** [comment]
OK let's get this in

**Pull request merged**